File: rest-service/src/main/java/mb/api/service/model/ErrorMap.java
package mb.api.service.model;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import mb.lib.blast.model.ToolOption;

/**
 * ErrorMap is a convenience implementation of the map structure used by
 * {@code UnprocessableEntityException}s.
 */
public final class ErrorMap extends HashMap<String, List<String>>
{
  /**
   * Constructs an empty {@code ErrorMap}.
   */
  public ErrorMap() {}

  public ErrorMap(int size) {
    super(size);
  }

  /**
   * Constructs an {@code ErrorMap} containing the given key mapped to a list
   * containing the given value.
   *
   * @param key   Error key.
   * @param value Error value.
   */
  public ErrorMap(String key, String value) {
    putError(key, value);
  }

  /**
   * Constructs an {@code ErrorMap} containing the given key mapped to a list
   * containing the given value.
   *
   * @param key   Error key.
   * @param value Error value.
   */
  public ErrorMap(ToolOption key, String value) {
    putError(key, value);
  }

  public ErrorMap putError(ToolOption key, String error) {
    putError(key.toString(), error);
    return this;
  }

  public ErrorMap putError(String key, String error) {
    computeIfAbsent(key, this::newList).add(error);
    return this;
  }

  public ErrorMap putError(String prefix, String key, String error) {
    computeIfAbsent(prefix + "." + key, this::newList).add(error);
    return this;
  }

  public ErrorMap putError(String prefix, String key, List<String> errors) {
    computeIfAbsent(prefix + "." + key, this::newList).addAll(errors);
    return this;
  }

  private List<String> newList(String ignored) {
    return new ArrayList<>();
  }
}

File: rest-service/src/main/kotlin/mb/api/service/model/ErrorMap.kt
package mb.api.service.model


import mb.lib.blast.model.ToolOption

/**
 * ErrorMap is a convenience implementation of the map structure used by
 * {@code UnprocessableEntityException}s.
 */
class ErrorMap: HashMap<String, MutableList<String>> {
  /**
   * Constructs an empty {@code ErrorMap}.
   */
  constructor(): super()

  constructor(size: Int): super(size)

  /**
   * Constructs an {@code ErrorMap} containing the given key mapped to a list
   * containing the given value.
   *
   * @param key   Error key.
   * @param value Error value.
   */
  constructor(key: String, value: String): super(1) {
    putError(key, value)
  }

  /**
   * Constructs an {@code ErrorMap} containing the given key mapped to a list
   * containing the given value.
   *
   * @param key   Error key.
   * @param value Error value.
   */
  constructor(key: ToolOption, value: String): super(1) {
    putError(key, value)
  }

  fun putError(key: ToolOption, error: String): ErrorMap {
    putError(key.toString(), error)
    return this
  }

  fun putError(key: String, error: String): ErrorMap {
    computeIfAbsent(key, this::newList).add(error)
    return this
  }

  fun putError(prefix: String, key: String, error: String): ErrorMap {
    computeIfAbsent(prefix + "." + key, this::newList).add(error)
    return this
  }

  fun putError(prefix: String, key: String, errors: List<String>): ErrorMap {
    computeIfAbsent("$prefix.$key", this::newList).addAll(errors)
    return this
  }

  private fun newList(ignored: String): MutableList<String> {
    return ArrayList()
  }
}
